FrEee/Scripts/AI/TestAI/MinistryOfInfrastructure.csx

```diff
@@ -0,0 +1,192 @@
+﻿#r "../../../bin/Debug/FrEee.Core.dll"
+#load "Plan.csx"
+using FrEee.Game.Objects.Civilization;
+using FrEee.Game.Objects.Space;
+using FrEee.Utility;
+using System;
+using System.Collections.Generic; 
+using System.Linq;
+using FrEee.Game.Objects.Orders;
+using FrEee.Game.Objects.Vehicles;
+using FrEee.Game.Interfaces;
+using FrEee.Utility.Extensions;
+using FrEee.Game.Objects.Commands;
+using FrEee.Game.Objects.Technology;
+
+
+/// <summary>
+/// Class devoted to the management of facilities.  
+/// </summary>
+public class MinistryOfInfrastructure
+{
+    public Empire Empire;
+
+    public Galaxy Galaxy;
+
+    public List<FacilityTemplate> AvailableFacilities; 
+
+
+    public void Run(Empire empire, Galaxy galaxy)
+    {
+        Empire = empire;
+        Galaxy = galaxy;
+        //get the latest versions of each facility. 
+        AvailableFacilities = Empire.Current.UnlockedItems.OfType<FacilityTemplate>()
+            .Where(f => f.ID == 0 && f.Cost.Any()).OnlyLatestVersions(f => f.Family).ToList();
+
+        ConstructNewFacilities();
+
+    }
+
+
+    /// <summary>
+    /// Construct facilities on any world where there are empty slots. 
+    /// </summary>
+    protected void ConstructNewFacilities()
+    {
+        var emptySlots = Empire.ColonizedPlanets.Where(x => x.Colony.ConstructionQueue.FacilitySlotsFree > 0); 
+
+        foreach(var planet in emptySlots)
+        {
+            var colony = planet.Colony; 
+            while(colony.ConstructionQueue.FacilitySlotsFree > 0)
+            {
+                var chosenFacility = GetEmpireVitalFaciltiesForSystem(planet.StarSystem);
+                if (chosenFacility == null)
+                    chosenFacility = GetPlanetVitalFacilities(planet);
+                if (chosenFacility == null)
+                    chosenFacility = GetResourceFacility(planet);
+                if (chosenFacility == null)
+                    continue; 
+                
+
+                var order = new ConstructionOrder<Facility, FacilityTemplate> { Template = chosenFacility };
+                colony.ConstructionQueue.Orders.Add(order);
+                var cmd = new AddOrderCommand
+                (
+                    colony.ConstructionQueue,
+                    order
+                );
+                Empire.Commands.Add(cmd);
+                Empire.LogAIMessage($"Construction of a {chosenFacility.Name} on {planet.Name} has been ordered");
+            }
+        }
+    }
+
+
+    /// <summary>
+    /// Ensure that all vital system options are taken care of. 
+    /// </summary>
+    /// <param name="starSystem"></param>
+    /// <returns></returns>
+    protected FacilityTemplate GetEmpireVitalFaciltiesForSystem(StarSystem starSystem)
+    {
+        //spaceports. 
+        var option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Spaceport"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Spaceport"))
+            return option;
```

Comment:
Should we skip this logic if the system is populated entirely by Natural Merchants population?